### PR TITLE
Upgrade loki and related dependencies

### PR DESCRIPTION
**Pull request by shantanualsi**
**What:**

This PR upgrades many packages - starting with loki, dskit, cortex, prometheus and even golang to their latest versions.

- The cortex project removes support for chunk storage, which is why the chunk-tool and corresponding packages - cassandra, filter, storage have to be removed from cortex-tools
- With the above deprecation, cortex also removes max_samples_per_query which has to be removed from overrides_exporter.go
- Lint errors - specifically usage of io/ioutil (deprecated in go 1.16) and gokit/kit/.. is fixed
- Go is upgraded to 1.21 to support all the upgraded modules

**Why:**
Users of cortex-tools require support for vector expressions that were recently introduced in Loki. This change will help cortex-tools support vector expressions at the expense of changes done as described above.

**Note for reviewer:**
The reason for a huge PR is that most of this cannot be broken down into multiple PRs due to go mod intertwined dependencies.

### CODE REVIEW

File: pkg/bench/ring_check.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ func NewRingChecker(id string, instanceName string, cfg RingCheckConfig, workloa
 		workload: workload,
 	}
 	reg := prometheus.DefaultRegisterer
-	cfg.MemberlistKV.MetricsRegisterer = reg
+	//cfg.MemberlistKV.MetricsRegisterer = reg
```
Comment (dannykopping) [inline pkg/bench/ring_check.go:58]: Curious why this was commented out?

File: pkg/commands/overrides_exporter.go [hunk 2]
```diff
@@ -108,9 +107,6 @@ func (o *OverridesExporterCommand) updatePresets(presetsMap map[string]*validati
 		o.presetsGauge.WithLabelValues(
 			"max_series_per_query", preset,
 		).Set(float64(limits.MaxSeriesPerQuery))
-		o.presetsGauge.WithLabelValues(
-			"max_samples_per_query", preset,
-		).Set(float64(limits.MaxSamplesPerQuery))
```
Comment (dannykopping) [inline pkg/commands/overrides_exporter.go:113]: Why was this removed?

**Review by dannykopping — COMMENTED** [review]
Firstly, a HUGE thanks for getting this done @shantanualsi! This was a massive effort.

LGTM, except there are a couple changes which I don't understand.

_If you do make any changes, please remember to not force-push; this makes incremental reviews impossible, and for huge PRs like this it's quite important_

File: pkg/commands/overrides_exporter.go (hunk 2 above)
Comment (shantanualsi) [reply pkg/commands/overrides_exporter.go:113]: This was removed in cortex while removing support for chunk storage (https://github.com/cortexproject/cortex/pull/4812/files#diff-1e0a07c0c1ea87ab5a582d79c587a73a9b8e0680d89c61d5c94be106823a4a4eL61) 
check `limits.go`

File: pkg/bench/ring_check.go (hunk 1 above)
Comment (shantanualsi) [reply pkg/bench/ring_check.go:58]: This was removed from dskit recently
https://github.com/grafana/dskit/pull/327/files

**Review by dannykopping — APPROVED** [review]
LGTM, thank you!

**Comment by shantanualsi** [comment]
Also, I forgot to mention yesterday, loki currently points to a branch of loki from main where I upgraded objstore. Please have a look. After merging the loki PR, I can point loki to `main` here too.

https://github.com/grafana/loki/pull/10451
